File: demo/flatbuffer_input_data/builtin_prefix/create_fb_data.py
import argparse
import flatbuffers

import os
import sys
sys.path.append("..")
sys.path.append("../..")

import convert_to_fb as tofb

def convert(from_file, to_file):
    with open(from_file, "r") as data_file:
        with open(to_file, "w+b") as out_f:
            builder = flatbuffers.Builder(1024)

            line_count = 0
            for line in data_file:
                line_count += 1
                simple_example = tofb.parse_simple(line)
                root_ex_obj = tofb.create_example_obj(builder, simple_example)
                
                # finish building and store to file
                builder.FinishSizePrefixed(root_ex_obj)
                buf = builder.Output()
                print(f"Writing buf of size {len(buf)}")
                out_f.write(buf)

                # create collection out of example
                root_coll_obj = tofb.create_example_collection_obj(builder, simple_example)
                # finish building and store to file
                builder.FinishSizePrefixed(root_coll_obj)
                buf = builder.Output()
                out_f.write(buf)

    print(
        f"Done converting [{line_count}] lines from text file [{from_file}] to flatbuffer file [{to_file}]"
    )


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--from-file",
        help="path to text file with simple examples in vw format",
        required=True,
    )
    parser.add_argument(
        "--to-file", help="path to flatbuffer output file", required=True
    )

    args = parser.parse_args()
    convert(args.from_file, args.to_file)


if __name__ == "__main__":
    main()

File: demo/flatbuffer_input_data/two_files/create_fb_data.cc
#include "convert_to_fb.h"

void convert(const std::string& from, const std::string& to)
{
  std::ifstream data_file(from);
  std::string line;

  std::ofstream outfile;
  outfile.open(to, std::ios::binary | std::ios::out);

  std::ofstream index_file;
  index_file.open(".index_file.txt");

  flatbuffers::FlatBufferBuilder builder;

  size_t line_count = 0;
  while (std::getline(data_file, line))
  {
    line_count++;
    auto simple_example = parse_line_simple(line);
    auto root_ex_obj = create_example_obj(builder, simple_example);

    builder.Finish(root_ex_obj);
    uint8_t* buf = builder.GetBufferPointer();
    size_t size = builder.GetSize();
    outfile.write(reinterpret_cast<char*>(buf), size);
    index_file << size << " " << static_cast<int>(object_type::example) << std::endl;

    auto root_coll_obj = create_example_collection_obj(builder, simple_example);

    builder.Finish(root_coll_obj);
    buf = builder.GetBufferPointer();
    size = builder.GetSize();
    outfile.write(reinterpret_cast<char*>(buf), size);
    index_file << size << " " << static_cast<int>(object_type::example_collection) << std::endl;
  }

  std::cout << "Done converting [" << line_count << "] lines from text file [" << from << "] to flatbuffer file [" << to
            << "]" << std::endl;
}

int main(int argc, char* argv[])
{
  if (argc < 3)
  {
    throw std::invalid_argument(
        "You need to provide two file names as arugments, the file to transform and the output file");
  }

  convert(argv[1], argv[2]);
  return 0;
}